### PR TITLE
Add random.shuffle_independent() and deprecate axis argument to shuffle()

### PR DESCRIPTION
**Pull request by jakevdp**
Addresses #2066

### CODE REVIEW

File: jax/random.py [hunk 1]
```diff
@@ -417,19 +418,66 @@ def _randint(key, shape, minval, maxval, dtype):
...
+def shuffle(key: np.ndarray, x: np.ndarray, axis: Optional[int]=0) -> np.ndarray:
+  """Shuffle the elements of an array uniformly along the first axis.
 
   Args:
     key: a PRNGKey used as the random key.
     x: the array to be shuffled.
-    axis: optional, an int axis along which to shuffle (default 0).
+    axis: (deprecated) an int axis along which to shuffle (default 0). If not None,
+      subarrays are shuffled independently along the specified axis, as in
+      :func:`~jax.random.shuffle_axis`. If `None`, subarrays are shuffled as a group.
```
Comment (shoyer) [inline jax/random.py:429]: should this be?
```suggestion
      :func:`~jax.random.shuffle_independent`. If `None`, subarrays are shuffled as a group.
```

File: jax/random.py [hunk 2]
```diff
@@ -417,19 +418,66 @@ def _randint(key, shape, minval, maxval, dtype):
...
+  if axis is None:
+    return x[_shuffle(key, np.arange(x.shape[0]), 0)]
+  if x.ndim != 1:
+    warnings.warn("shuffle(): int axis with multi-dimenstional input is deprecated. "
+      "Set axis=None to sort groups along the first axis, or use shuffle_independent() "
+      "to sort subarrays independently along a specified axis.", JaxDeprecationWarning)
   return _shuffle(key, x, axis)
 
+
+def shuffle_independent(key: np.ndarray, x: np.ndarray, axis: int=0) -> np.ndarray:
```
Comment (shoyer) [inline jax/random.py:455]: Should this be `random.permutation`, for consistency with NumPy?

File: jax/random.py [hunk 3]
```diff
@@ -417,19 +418,66 @@ def _randint(key, shape, minval, maxval, dtype):
...
   Args:
     key: a PRNGKey used as the random key.
     x: the array to be shuffled.
-    axis: optional, an int axis along which to shuffle (default 0).
+    axis: (deprecated) an int axis along which to shuffle (default 0). If not None,
+      subarrays are shuffled independently along the specified axis, as in
+      :func:`~jax.random.shuffle_axis`. If `None`, subarrays are shuffled as a group.
+      along the leading axis, as in :func:`numpy.random.shuffle`.
+      In a future release, the default will be changed to `None` and support for
+      integer axis values will be dropped.
```
Comment (shoyer) [inline jax/random.py:432]: Long term, do we need support for shuffling in groups at all?

- NumPy seems to get along without it OK, just creating explicit arrays to use with indexing
- We have `vmap`, so arguably the idiomatic way to shuffle a bunch of arrays together is to `vmap` but not batch the key.

File: jax/util.py [hunk 4]
```diff
@@ -230,3 +230,7 @@ def wrap_name(name, transform_name):
 
 def extend_name_stack(stack, name=''):
   return stack + name + '/'
+
+
+class JaxDeprecationWarning(UserWarning):
+  pass
```
Comment (shoyer) [inline jax/util.py:236]: is there a reason why we should use the builtin DeprecationWarning or FutureWarning?

File: jax/random.py (hunk 2 above)
Comment (jakevdp) [reply jax/random.py:455]: Done

Comment (jakevdp) [reply jax/random.py:455]: Oops, wrong comment. I think permutation is different... numpy doesn't have any function to shuffle each subarray independently along an axis. I suppose an alternative would be to drop this functionality entirely.

File: jax/util.py (hunk 4 above)
Comment (jakevdp) [reply jax/util.py:236]: I avoided DeprecationWarning because it is hidden by default, so not very useful to users. FutureWarning would be fine, though.

**Comment by jakevdp** [comment]
Closing this in favor of a simpler approach: https://github.com/google/jax/issues/2066#issuecomment-621216589
